File: app/javascript/channels/room_channel.js
import consumer from "./consumer";

const chatChannel = consumer.subscriptions.create("RoomChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  // room_channel.rbでブロードキャストされたものがここに届く
  // 非同期でアラートの代わりに画面に文字を表示させる
  received: function (data) {
    return $("#messages").append(data["message"]);
    // Called when there's incoming data on the websocket for this channel
  },

  // channelのspeakを実行させるためにjsでspeak関数を定義する
  // これが実行されるとコンシューマになったRoomChannel#speak({ message: message })が呼ばれる
  speak: function (message) {
    return this.perform("speak", {
      message: message,
    });
  },
});

$(document).on("keypress", "[data-behavior~=room_speaker]", function (event) {
  if (event.keyCode === 13) {
    chatChannel.speak(event.target.value);
    event.target.value = "";
    return event.preventDefault();
  }
});

File: app/channels/room_channel.rb
class RoomChannel < ApplicationCable::Channel
  def subscribed
    # room_channelをストリームする
    stream_from "room_channel"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak(data)
    # jsで実行されたspeakのmessageを受け取り、room_channelのreceivedにブロードキャストする
    # ActionCable.server.broadcast 'room_channel', message: data['message']

    # speakアクションの書き換え
    Message.create! content: data['message']
  end
end
